src/source/src/dml_manager.rs

```diff
@@ -32,35 +34,63 @@ pub type DmlManagerRef = Arc<DmlManager>;
 /// channel instead of offering a `write_chunk` interface).
 #[derive(Default, Debug)]
 pub struct DmlManager {
-    batch_dmls: Mutex<HashMap<TableId, TableSourceRef>>,
+    batch_dmls: RwLock<HashMap<TableId, Weak<TableSource>>>,
 }
 
 impl DmlManager {
     pub fn new() -> Self {
         Self {
-            batch_dmls: Mutex::new(HashMap::new()),
+            batch_dmls: RwLock::new(HashMap::new()),
         }
     }
 
     pub fn register_reader(
```

Comment:
According to the method name, how about directly returning the `TableStreamReader`?

Reply:
If we directly return `TableStreamReader` here, then no one will get hold of the strong ref of `TableSource`. It will thus be dropped immediately because only `DmlManager` has a weak ref of it. We might have to refactor `TableSource` later.

Reply:
Yes. Makes sense.